File: tell-dont-ask-kata-kotlin/src/test/java/it/gabrieletondi/telldontaskkata/useCase/OrderCreationUseCaseTest.kt
package it.gabrieletondi.telldontaskkata.useCase


import it.gabrieletondi.telldontaskkata.domain.Category
import it.gabrieletondi.telldontaskkata.domain.OrderStatus
import it.gabrieletondi.telldontaskkata.domain.Product
import org.junit.Test

import java.math.BigDecimal
import java.util.ArrayList
import java.util.Arrays

import org.hamcrest.Matchers.hasSize
import org.hamcrest.Matchers.`is`
import org.junit.Assert.assertThat
import it.gabrieletondi.telldontaskkata.doubles.InMemoryProductCatalog
import it.gabrieletondi.telldontaskkata.doubles.TestOrderRepository

class OrderCreationUseCaseTest {
    private val orderRepository = TestOrderRepository()
    private val food = object : Category() {
        init {
            name = "food"
            taxPercentage = BigDecimal("10")
        }
    }
    private val productCatalog = InMemoryProductCatalog(
        Arrays.asList(
            object : Product() {
                init {
                    name = "salad"
                    price = BigDecimal("3.56")
                    category = food
                }
            },
            object : Product() {
                init {
                    name = "tomato"
                    price = BigDecimal("4.65")
                    category = food
                }
            }
        )
    )
    private val useCase = OrderCreationUseCase(orderRepository, productCatalog)

    @Test
    @Throws(Exception::class)
    fun sellMultipleItems() {
        val saladRequest = SellItemRequest()
        saladRequest.productName = "salad"
        saladRequest.quantity = 2

        val tomatoRequest = SellItemRequest()
        tomatoRequest.productName = "tomato"
        tomatoRequest.quantity = 3

        val request = SellItemsRequest()

        (request.requests as ArrayList<SellItemRequest>).add(saladRequest)
        (request.requests as ArrayList<SellItemRequest>).add(tomatoRequest)

        useCase.run(request)

        val insertedOrder = orderRepository.savedOrder
        assertThat(insertedOrder!!.status, `is`(OrderStatus.CREATED))
        assertThat(insertedOrder.total, `is`(BigDecimal("23.20")))
        assertThat(insertedOrder.tax, `is`(BigDecimal("2.13")))
        assertThat(insertedOrder.currency, `is`("EUR"))
        assertThat(insertedOrder.items, hasSize(2))
        assertThat(insertedOrder.items!![0].product!!.name, `is`("salad"))
        assertThat(insertedOrder.items!![0].product!!.price, `is`(BigDecimal("3.56")))
        assertThat(insertedOrder.items!![0].quantity, `is`(2))
        assertThat(insertedOrder.items!![0].taxedAmount, `is`(BigDecimal("7.84")))
        assertThat(insertedOrder.items!![0].tax, `is`(BigDecimal("0.72")))
        assertThat(insertedOrder.items!![1].product!!.name, `is`("tomato"))
        assertThat(insertedOrder.items!![1].product!!.price, `is`(BigDecimal("4.65")))
        assertThat(insertedOrder.items!![1].quantity, `is`(3))
        assertThat(insertedOrder.items!![1].taxedAmount, `is`(BigDecimal("15.36")))
        assertThat(insertedOrder.items!![1].tax, `is`(BigDecimal("1.41")))
    }

    @Test(expected = UnknownProductException::class)
    @Throws(Exception::class)
    fun unknownProduct() {
        val request = SellItemsRequest()
        request.requests = ArrayList()
        val unknownProductRequest = SellItemRequest()
        unknownProductRequest.productName = "unknown product"
        (request.requests as ArrayList<SellItemRequest>).add(unknownProductRequest)

        useCase.run(request)
    }
}

File: tell-dont-ask-kata-csharp/TellDontAsk.Tests/useCase/OrderCreationUseCaseTest.cs
using System.Collections.Generic;
using Xunit;

public class OrderCreationUseCaseTest
{
    private readonly TestOrderRepository orderRepository;

    private readonly Category food;

    private readonly ProductCatalog productCatalog;
    private readonly OrderCreationUseCase useCase;

    public OrderCreationUseCaseTest()
    {
        orderRepository = new TestOrderRepository();

        food = new Category();
        food.SetName("food");
        food.SetTaxPercentage(10m);

        var salad = new Product();
        salad.SetName("salad");
        salad.SetPrice(3.56m);
        salad.SetCategory(food);

        var tomato = new Product();
        tomato.SetName("tomato");
        tomato.SetPrice(4.65m);
        tomato.SetCategory(food);

        productCatalog = new InMemoryProductCatalog(new List<Product>() { salad, tomato });
        useCase = new OrderCreationUseCase(orderRepository, productCatalog);
    }


    [Fact]
    public void SellMultipleItems()
    {
        SellItemRequest saladRequest = new SellItemRequest();
        saladRequest.SetProductName("salad");
        saladRequest.SetQuantity(2);

        SellItemRequest tomatoRequest = new SellItemRequest();
        tomatoRequest.SetProductName("tomato");
        tomatoRequest.SetQuantity(3);

        SellItemsRequest request = new SellItemsRequest();
        request.SetRequests(new List<SellItemRequest>());
        request.GetRequests().Add(saladRequest);
        request.GetRequests().Add(tomatoRequest);

        useCase.Run(request);

        Order insertedOrder = orderRepository.GetSavedOrder();
        Assert.Equal(OrderStatus.CREATED, insertedOrder.GetStatus());
        Assert.Equal(23.20m, insertedOrder.GetTotal());
        Assert.Equal(2.13m, insertedOrder.GetTax());
        Assert.Equal("EUR", insertedOrder.GetCurrency());
        Assert.Equal(2, insertedOrder.GetItems().Count);
        Assert.Equal("salad", insertedOrder.GetItems()[0].GetProduct().GetName());
        Assert.Equal(3.56m, insertedOrder.GetItems()[0].GetProduct().GetPrice());
        Assert.Equal(2, insertedOrder.GetItems()[0].GetQuantity());
        Assert.Equal(7.84m, insertedOrder.GetItems()[0].GetTaxedAmount());
        Assert.Equal(0.72m, insertedOrder.GetItems()[0].GetTax());
        Assert.Equal("tomato", insertedOrder.GetItems()[1].GetProduct().GetName());
        Assert.Equal(4.65m, insertedOrder.GetItems()[1].GetProduct().GetPrice());
        Assert.Equal(3, insertedOrder.GetItems()[1].GetQuantity());
        Assert.Equal(15.36m, insertedOrder.GetItems()[1].GetTaxedAmount());
        Assert.Equal(1.41m, insertedOrder.GetItems()[1].GetTax());
    }

    [Fact]
    public void UnknownProduct()
    {
        SellItemsRequest request = new SellItemsRequest();
        request.SetRequests(new List<SellItemRequest>());
        SellItemRequest unknownProductRequest = new SellItemRequest();
        unknownProductRequest.SetProductName("unknown product");
        request.GetRequests().Add(unknownProductRequest);

        Assert.Throws<UnknownProductException>(() =>useCase.Run(request));
    }
}